piqe_ocp_lib/api/resources/ocp_machine_management.py

```diff
@@ -353,3 +353,120 @@ def is_machine_deleted(self, machine_name: str, timeout: int = TIMEOUT) -> bool:
+            mhc_instance_del = self.machinehealthcheck.delete(
+                name=machine_health_check_name, field_selector=field_selector, namespace="openshift-machine-api"
+            )
+        except ApiException as e:
+            logger.error("Exception while deleting Machine Health Check: %s\n", e)
+            raise Exception(e)
+        logger.info(f"Deleted MachineHealthCheck: {machine_health_check_name}")
+        return mhc_instance_del
+
+    def is_machine_health_check_configured(self, machine_set_name: str, machine_health_check_name: str) -> bool:
+        """
+        if machine health check is configured on particular machineset, the status would be updated
+        """
+        field_selector = f"metadata.name={machine_health_check_name}"
+        TIMEOUT = 300
+        try:
+            for event in self.machinehealthcheck.watch(
+                namespace="openshift-machine-api", field_selector=field_selector, timeout=TIMEOUT
+            ):
+                currentHealthy = event["raw_object"]["status"]["currentHealthy"]
+                if currentHealthy != 0:
+                    logger.info(
+                        f"{currentHealthy} machines in Machineset: {machine_set_name} \
+                            configured with machine health check"
+                    )
+                    return True
+        except ApiException as e:
+            logger.error("Exception while watching events of Machine Health Check: %s\n", e)
+            raise Exception(e)
+
+    def configure_machine_health_check(self, machine_set_name: str, machine_health_check_name: str) -> bool:
+        """
+        Create Machine Health Check for a machineset.
+        :param machine_health_check_name: (str) name of the machine health check
+        :return: (bool) True when successfully configured a Machine Heath Check object OR False otherwise
+        """
+        mhc_data = {
+            "kind": "MachineHealthCheck",
+            "spec": {
+                "unhealthyConditions": [
+                    {"status": "False", "type": "Ready", "timeout": "300s"},
+                    {"status": "Unknown", "type": "Ready", "timeout": "300s"},
+                ],
+                "maxUnhealthy": "40%",
+                "nodeStartupTimeout": "10m",
+                "selector": {"matchLabels": {"machine.openshift.io/cluster-api-machineset": f"{machine_set_name}"}},
+            },
+            "apiVersion": "machine.openshift.io/v1beta1",
+            "metadata": {"namespace": "openshift-machine-api", "name": f"{machine_health_check_name}"},
+        }
+
+        self.create_machine_health_check(machine_set_name, machine_health_check_name, mhc_data)
+        if not self.is_machine_health_check_configured(machine_set_name, machine_health_check_name):
+            raise AssertionError("The Machine Health Check is not configured as none of the machine are healthy")
+        self.delete_machine_health_check(machine_health_check_name)
```

Comment:
Can you please elaborate on the distinction between the create and configure? the way I am reading it is that create is a raw PUT operation where we pass the whole body, then configure is a more generic version where most of the body is already predefined and we just get to assign the machine set name machine health check names, correct? My main confusion is on the 'delete' call in line 471. If I am following this right, the watch method in line 469 will either return True or raise an exception, would we ever execute line 470? and so, is the intent of this method to always to perform a delete at the end? I initially thought you meant to delete if the watch methods fails but from what I see, the indentation of the delete method is the same as the if statement.

Reply:
> Can you please elaborate on the distinction between the create and configure? the way I am reading it is that create is a raw PUT operation where we pass the whole body, then configure is a more generic version where most of the body is already predefined and we just get to assign the machine set name machine health check names, correct?

This is correct I think I understand why this has become confusing. 

I have now moved the configure from api to test as a private function. We would now have better distinction between create and configure. As configure is kind of a helper function which would call create, watch, delete func from the resource api.

> My main confusion is on the 'delete' call in line 471. If I am following this right, the watch method in line 469 will either return True or raise an exception, would we ever execute line 470? and so, is the intent of this method to always to perform a delete at the end? I initially thought you meant to delete if the watch methods fails but from what I see, the indentation of the delete method is the same as the if statement.

Right delete has to run always irrespective of the what is_machine_health_check_configured returns because we have created a new mhc_instance it should be deleted after the check is performed.

Have wrapped the function into a try except & finally block & moved delete to finally block.

have added another test to check the false condition as well.

Following is the run for the new snippet:
```
[root@96106851a09b piqe-ocp-lib]# pytest -sv piqe_ocp_lib/tests/resources/test_ocp_machine_management.py::TestOcpMachineHealthCheck
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.9.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
OpenShift version: latest
rootdir: /opt/src/piqe-ocp-lib, configfile: pyproject.toml
collected 4 items                                                                                                                                                          

piqe_ocp_lib/tests/resources/test_ocp_machine_management.py::TestOcpMachineHealthCheck::test_get_all_machine_health_checks 2021-09-30 09:44:31,178 INFO (log_start_test_case) Starting test case: test_get_all_machine_health_checks
PASSED2021-09-30 09:44:31,822 INFO (log_end_test_case) Ending test case: test_get_all_machine_health_checks

piqe_ocp_lib/tests/resources/test_ocp_machine_management.py::TestOcpMachineHealthCheck::test_get_machine_health_check 2021-09-30 09:44:31,827 INFO (log_start_test_case) Starting test case: test_get_machine_health_check
PASSED2021-09-30 09:44:32,851 INFO (log_end_test_case) Ending test case: test_get_machine_health_check

piqe_ocp_lib/tests/resources/test_ocp_machine_management.py::TestOcpMachineHealthCheck::test_machine_health_check_is_configured 2021-09-30 09:44:32,854 INFO (log_start_test_case) Starting test case: test_machine_health_check_is_configured
2021-09-30 09:44:33,936 - [INFO] - piqe_api_logger - ocp_machine_management@create_machine_health_check:412 - MachineHealthCheck: new-mhc created for Machineset:dell-r730-068-0-r62wl-worker-0
2021-09-30 09:44:34,265 - [INFO] - piqe_api_logger - ocp_machine_management@is_machine_health_check_configured:439 - 1 machines in Machineset: dell-r730-068-0-r62wl-worker-0                             configured with machine health check
2021-09-30 09:44:35,442 - [INFO] - piqe_api_logger - ocp_machine_management@delete_machine_health_check:424 - Deleted MachineHealthCheck: new-mhc
PASSED2021-09-30 09:44:35,445 INFO (log_end_test_case) Ending test case: test_machine_health_check_is_configured

piqe_ocp_lib/tests/resources/test_ocp_machine_management.py::TestOcpMachineHealthCheck::test_machine_health_check_is_not_configured 2021-09-30 09:44:35,448 INFO (log_start_test_case) Starting test case: test_machine_health_check_is_not_configured
2021-09-30 09:44:36,340 - [INFO] - piqe_api_logger - ocp_machine_management@create_machine_health_check:412 - MachineHealthCheck: new-mhc created for Machineset:wrong_machine_set_name
2021-09-30 09:44:36,681 - [ERROR] - piqe_api_logger - ocp_machine_management@is_machine_health_check_configured:444 - None machines in Machinset: wrong_machine_set_name configured
2021-09-30 09:44:36,683 - [ERROR] - piqe_api_logger - test_ocp_machine_management@_configure_machine_health_check:218 - The Machine Health Check is not configured as none of the machines are healthy
2021-09-30 09:44:37,771 - [INFO] - piqe_api_logger - ocp_machine_management@delete_machine_health_check:424 - Deleted MachineHealthCheck: new-mhc
PASSED2021-09-30 09:44:37,774 INFO (log_end_test_case) Ending test case: test_machine_health_check_is_not_configured


============================================================================ 4 passed in 29.73s ============================================================================

```

Reply:
Thank you for the clarification @bhavikbhavsar